### PR TITLE
522 chat click links

### LINKED ISSUE #522: clickable links in chat
extracted from https://github.com/Qabel/qabel-desktop/issues/519

Text is now copyable (as soon as #519 is implemented), so links can be copied into the browser, but clickable links would be far better in terms of usability. ~~Unfortunately, that feature is harder to implement than just selecting text (which was achieved by replacing the Label by a styled TextArea) so it was split out of the original ticket.~~

(needs to be discussed in terms of future enhancements ... images, code, citations, etc.)

### PR DESCRIPTION
**Pull request by nnice**
resolves #522

### CODE REVIEW

File: src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/PlaintextMessageRenderer.java [hunk 1]
```diff
@@ -2,20 +2,73 @@
...
+        String text = renderString(dropPayload, resourceBundle);
+        return renderHyperlinks(text);
+    }
+
+    @NotNull
+    TextFlow renderHyperlinks(String message) {
+        String regex = "(https?:\\/\\/(?:www\\.|(?!www))[^\\s\\.]+\\.[^\\s]{2,}|www\\.[^\\s]+\\.[^\\s]{2,})";
+        Pattern pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE | Pattern.COMMENTS | Pattern.MULTILINE);
+        Matcher matcher = pattern.matcher(message);
+        String result = matcher.replaceAll("[$1]");
```
Comment (audax) [inline src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/PlaintextMessageRenderer.java]: Extract this replace stuff to a diffirent place, it shouldn't be in a renderer.

File: src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/PlaintextMessageRenderer.java [hunk 2]
```diff
@@ -2,20 +2,73 @@
...
+import java.util.function.Consumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public class PlaintextMessageRenderer implements FXMessageRenderer {
+    private static final String STYLE_CLASS = "message-text";
+    public Consumer<String> browserOpener;
+
+    public PlaintextMessageRenderer() {
+        browserOpener = (uri) -> {
```
Comment (audax) [inline src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/PlaintextMessageRenderer.java]: You can also initialize it inline where you declare the member. Functionally it is identical.

A better way do to this is imho to create a dedicated browser opener interactor and replace the real one, which has Desktop.getDesktop() stuff, with a testable browser opener.

File: src/main/resources/main.css [hunk 3]
```diff
@@ -155,14 +155,14 @@
...
-}
-
 #myMessagebox {
     -fx-background-color: #4d4d4d;
     -fx-background-radius: 5;
 }
+#shadow {
+    -fx-effect: dropshadow(one-pass-box, grey, 5, -1, 0, 2);
+}
+
```
Comment (julianseeger) [inline src/main/resources/main.css]: test

**Comment by julianseeger** [comment]
what's the state here?
Linewrapping doesn't work anymore as far as I can tell:
image

**Comment by nnice** [comment]
gonna check this

File: src/main/java/de/qabel/desktop/DesktopClientGui.java [hunk 4]
```diff
@@ -118,7 +118,7 @@ private void showLayoutStage() {
         Scene layoutScene = new Scene(view, 900, 600, true, SceneAntialiasing.BALANCED);
         Platform.runLater(() -> primaryStage.setScene(layoutScene));
         primaryStage.show();
-        closeStage();
+//        closeStage();
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClientGui.java]: needs to be activated again

File: src/main/java/de/qabel/desktop/ui/actionlog/util/QabelChatSkin.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import javafx.scene.Node;
+import javafx.scene.control.Control;
+import javafx.scene.text.Text;
+import javafx.scene.text.TextFlow;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+class QabelChatSkin extends BehaviorSkinBase<QabelChatLabel, BehaviorBase<QabelChatLabel>> {
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/actionlog/util/QabelChatSkin.java]: remove this legacy code

File: src/main/resources/main.css [hunk 6]
```diff
@@ -458,28 +486,6 @@ VBox.spaced, HBox.spaced {
     -fx-border-width: 0 0 1px 0;
 }
 
-.hyperlink {
```
Comment (julianseeger) [inline src/main/resources/main.css]: where has this been used before? seems like there are links in the gui now that are no more visible as links...

**Comment by julianseeger** [comment]
feedback from po for the links:
 * every scheme should be detected (like ftp://example.org)
 * every url without a scheme should be detected IF it has a valid TLD

**Comment by julianseeger** [comment]
and the new message indicator does never disappear with this code

File: src/main/resources/main.css (hunk 6 above)
Comment (julianseeger) [inline src/main/resources/main.css]: was in the about screen but isn't used anymore

**Comment by julianseeger** [comment]
image

and it only took two regular expressions...

File: src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/partial/HyperlinkRenderer.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        return IP_PATTERN.matcher(uri.getHost()).matches();
+    }
+
+    private void checkScheme(String match, URI uri) throws URISyntaxException {
+        if (uri.getScheme() == null) {
+            throw new URISyntaxException(match, "missing scheme");
+        }
+    }
+
+    private boolean hasValidTld(URI uri) throws URISyntaxException {
```
Comment (audax) [inline src/main/java/de/qabel/desktop/ui/actionlog/item/renderer/partial/HyperlinkRenderer.java]: I really think that filtering out unknown TLDs is not a good idea.

**Pull request merged**